repository/src/main/java/com/epam/esm/repository/TagRepository.java

```diff
@@ -0,0 +1,26 @@
+package com.epam.esm.repository;
+
+import com.epam.esm.model.Tag;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+@Repository
+public interface TagRepository extends JpaRepository<Tag,Long> {
+    Set<Tag> findAllByNameIn(List<String> tagNames);
+    @Query(value = "select t.id, t.name " +
+            "from orders o " +
+            "         join users u on o.user_id = u.id" +
+            "         join gift_certificate g on g.id = o.gift_certificate_id" +
+            "         join gift_certificate_tag gct on g.id = gct.gift_id" +
+            "         join tag t on t.id = gct.tag_id " +
+            "where u.id = ?1 " +
+            "group by t.name, t.id " +
+            "order by sum(o.price) desc " +
+            "limit 1", nativeQuery = true)
+    Optional<Tag> findMostUsedTagByUser(Long userId);
```

Comment:
Is it written to get the most widely used tag of a user with the highest cost of all orders? Here you are using limit 1 and ordering by price desc. It doesn't get the most widely used tag, it gets the most expensive one, even if there is one tag there. And you added limit 1. It means that if there are two tags which are used at the same time, and their prices are the same, this query gets only one of them. I think this query can not give the required functionality.
Imagine there are these tables:
Tag table
id - 1, name - tech
id - 2, name - book
id -3, name - gym
 
Gift certificate table
id - 1, price - 1000, tagId - 1
id - 2, price - 2000, tagId - 1
id - 3, price - 3000, tagId - 1
id - 4, price - 3000, tagId - 2
id - 5, price - 2000, tagId - 2
id - 6, price - 1000, tagId - 2
id - 7, price - 4000, tagId - 3
 
orders
id - 1, cost - 1000, certificateId - 1, userId - 1
id - 2, cost - 2000, certificateId - 2, userId - 1
id - 3, cost - 3000, certificateId - 3, userId - 1
id - 4, cost - 3000, certificateId - 4, userId - 1
id - 5, cost - 2000, certificateId - 5, userId - 1
id - 6, cost - 1000, certificateId - 6, userId - 1
id - 7, cost - 4000, certificateId - 7, userId - 1

User with id 1, bought 7 certificates. The most widely used tags are tech and book, three times. Their prices are the same, 6000. So both of them should be returned.